src/build/build.ts

```diff
@@ -77,9 +75,9 @@ export function build(options: BuildOptions, config: ProjectConfig): Promise<any
     logger.debug(`Reading source files...`);
     let sourcesStream = polymerProject.sources()
       .pipe(polymerProject.splitHtml())
-      .pipe(gulpif(/\.js$/, uglify(optimizeOptions.js)))
-      .pipe(gulpif(/\.css$/, cssSlam(optimizeOptions.css)))
-      .pipe(gulpif(/\.html$/, htmlmin(optimizeOptions.html)))
+      .pipe(gulpif(/\.js$/, new JSOptimizeStream(optimizeOptions.js)))
```

Comment:
It's not going to work :( And I really wanted it to.

There are two big problems in the way:

1. Lots of streams end up emitting an error when one is encountered. The only thing that is sent with the event is an error object. This means we have no reliable way to know which file failed, which we need to know so that we can pass through the original. No level of monkey patching can generally solve for this.
2. There is no consistency with how these streams are implemented. Some return an error to the `_transform` callback, some weren't even written with the `_transform` callback at all and instead rely on the `_write` callback, some do neither and just emit an error, some (like gulp-html-minifier') just throw an error, etc. etc. etc. Because of all the different changes Node streams have gone through over the years, a lot of these are all valid implementations. So even if #1 wasn't an issue, it would still be incredibly hard to create a general wrapper.

BUT what I have here is almost exactly what we want. Instead of wrapping a stream it wraps the actual libraries we want to use. This grants the user wayyyy more flexibility into how the library is called and how its options are consumed. Check out the implementation of each of these, and you'll see they're just small wrappers around `GenericOptimizeStream`. This solution also has a smaller dependency footprint (1 less level of dependencies, see [html-minifier](http://npm.anvaka.com/#/view/2d/html-minifier) vs. [gulp-html-minifier](http://npm.anvaka.com/#/view/2d/gulp-html-minifier))

WDYT? If we like, we can create a factory similar to `passThroughOnFail` for `GenericOptimizeStream` (takes its constructor arguments and returns a new stream).